contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -88,31 +97,56 @@ pub fn call_contract(
         ),
     )
     .change_context(Error::InvalidCrossChainId)?;
-    let source_address = Address::from_str(sender.as_str())
-        .change_context(Error::InvalidSourceAddress(sender.clone()))?;
+    let source_address = Address::from_str(info.sender.as_str())
+        .change_context(Error::InvalidSourceAddress(info.sender))?;
     let msg = call_contract.to_message(id, source_address);
 
     state::save_unique_routable_msg(storage, &msg.cc_id, &msg)
         .inspect_err(|err| panic_if_already_exists(err, &msg.cc_id))
         .change_context(Error::SaveRoutableMessage)?;
 
-    Ok(
-        route_to_router(storage, &Router::new(router), vec![msg.clone()])?.add_event(
-            AxelarnetGatewayEvent::ContractCalled {
-                msg,
+    let res = match info.funds.as_slice() {
```

Comment:
Use `info.funds.pop()` to avoid cloning the token

Reply:
That would require `info` to be a mutable reference which I strongly do not like.

Reply:
The helper can take `info.funds` by value, and mark it as a mutable arg since it owns it, so it's an implementation detail, and doesn't require `info` to marked as mutable.

Reply:
It doesn't make sense for it to own it though. Other code may need the `info` for a completely different reason.

Reply:
Did it but it isn't anything cleaner.

Reply:
you can move just `info.funds` instead of all of `info`. But not a strong opinion

Reply:
I think I like what I did better, i.e. implemented a method on `MessageInfo`.